### PR TITLE
Fix a segfault I get in a complicated solver.

### PR DESCRIPTION
**Pull request by pefarrell**
Debugged with Lawrence. PETSc's handling of how to assemble problems via field splits and coarsenings is such a mess.

### CODE REVIEW

**Comment by pefarrell** [comment]
It's not easy, no. The exhibiting code depends on a library. It would be some work to extract it.

**Comment by wence-** [comment]
The proximate issue is MG + fieldsplit + aux operator on a mixed space and then a split inside? So can you provoke it by just doing a 3-way mass-matrix on two levels and splitting 2 fields out and then splitting inside that?

**Comment by pefarrell** [comment]
OK, test added. It segfaults (on my machine) without this PR, but runs successfully with it.

**Comment by pefarrell** [comment]
(The test was prepared by Ioannis Papadopoulos.)

File: firedrake/dmhooks.py [hunk 1]
```diff
@@ -429,6 +429,7 @@ def coarsen(dm, comm):
                  teardown=partial(pop_appctx, cdm, cctx),
                  call_setup=True)
         # Necessary for MG inside a fieldsplit in a SNES.
+        dm.setKSPComputeOperators(firedrake.solving_utils._SNESContext.compute_operators)
```
Comment (colinjcotter) [inline firedrake/dmhooks.py:432]: This is a bit mysterious! Can you add a comment explaining why?

Comment (wence-) [reply firedrake/dmhooks.py:432]: I'll add it in post. Basically there are hooks that are performed by petsc that attempt to carry compute operators around from outer to inner dms. The outer DM has a kspcomputeoperators_snes attached to it (which is wrong inside a fieldsplit for complex reasons) and that gets transferred onto the coarse DM after we've set up our compute operators. This way we make sure the fine DM has the right compute operators callback.

**Pull request merged**